src/javascript/html-template-literal-scanner.ts

```diff
@@ -0,0 +1,104 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as babel from 'babel-types';
+
+import {ScannedInlineDocument} from '../model/model';
+
+import {getIdentifierName} from './ast-value';
+import {Visitor} from './estree-visitor';
+import {JavaScriptDocument} from './javascript-document';
+import {JavaScriptScanner} from './javascript-scanner';
+
+export class InlineHtmlDocumentScanner implements JavaScriptScanner {
+  async scan(
+      document: JavaScriptDocument,
+      visit: (visitor: Visitor) => Promise<void>) {
+    const features: ScannedInlineDocument[] = [];
+
+    const myVisitor: Visitor = {
+      enterTaggedTemplateExpression(node) {
+        const inlineDocument = getInlineDocument(node, document);
+        if (inlineDocument !== undefined) {
+          features.push(inlineDocument);
+        }
+      }
+    };
+
+    await visit(myVisitor);
+
+    return {features};
+  }
+}
+
+export function getInlineDocument(
+    node: babel.TaggedTemplateExpression,
+    parsedDocument: JavaScriptDocument): ScannedInlineDocument|undefined {
+  const tagName = getIdentifierName(node.tag);
+  if (tagName === undefined || !/(^|\.)html$/.test(tagName)) {
+    return;
+  }
+
+  const sourceRangeForLiteral = parsedDocument.sourceRangeForNode(node.quasi);
+  if (sourceRangeForLiteral === undefined) {
+    return;
+  }
+  const sourceRangeForContents = {
+    file: sourceRangeForLiteral.file,
+    start: {
+      line: sourceRangeForLiteral.start.line,
+      column: sourceRangeForLiteral.start.column + 1
+    },
+    end: {
+      line: sourceRangeForLiteral.end.line,
+      column: sourceRangeForLiteral.end.column - 1
+    }
+  };
+
+  let contents = '';
+  let previousEnd: number|undefined;
+  for (let i = 0; i < node.quasi.quasis.length; i++) {
+    const quasi = node.quasi.quasis[i];
+    if (previousEnd !== undefined) {
+      const fullExpressionTextWithDelimitors =
+          parsedDocument.contents.slice(previousEnd, quasi.start);
+      // Replace everything but whitespace in ${expressions} (including the
```

Comment:
I don't quite understand this. Can't we just skip the whole expression? We're assuming well-formed HTML in the literal parts (which should be documented). I'm not sure we can put anything useful in the contents.

Reply:
Why do you need to replace with space or newline, and even so, couldn't this be a map over the quasis?

```ts
const contents = node.quasi.quasis.join('');
```

Reply:
Consider the js file:

```js
html`<div>${
    'Hello world'
  }</div>
```

If we remove the expression entirely, the html parser receives `<div></div>` and when we ask for the source range of the closing tag it'll give one on the first line, and starting just after the `<div>`. By preserving newlines and replacing every other character with a space, the source map will correctly be able to tell it starts on the third line, fourth character.

Reply:
Got it. Add that explanation to the comments ;)

Reply:
Done